### DISCUSSION TITLE
Per-object collision filter

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#215] [Q&A]
Is there (or could there be) a good way to disable collisions between two specific objects? We've been doing this in several cases, like when an object attaches itself to another one. It was easier to have both objects keep their physics bodies instead of merging them into a combined one, and potentially separating them again later.

**Comment by jrouwe** [comment]
Yes, this is what the GroupFilter virtual interface is for. Each body has a CollisionGroup (which contains a GroupFilter reference). The Collision group has  64 bits of custom data and you can use that in the GroupFilter::CanCollide function to determine if two objects should collide. An example implementation is GroupFilterTable and an example of how to use it in the case of ragdolls can be found in RagdollSettings::DisableParentChildCollisions.

**Reply by DocAce** [reply]
I've already been using the GroupID for categorization of objects (e.g. one group for asteroids, another one for drops, etc) to broadly filter various types of collisions, mainly but not only for performance reasons. The given example implementation is very nice but assumes a "reasonable" number of sub groups. Unfortunately I cannot set any reasonable limit here, and I also have to support filtering across Groups (e.g. a certain ship may want to disable collisions with a certain asteroid as it's picking it up). I'm exploring an alternative method right now that puts the BodyID as the SubGroup, but it'll require looking up data in a map during collision detection, which I had hoped to be able to avoid. Still, it'll work and be mostly fine as we've done that previously when we were using Bullet.

**Reply by jrouwe** [reply]
What we do is to have an auto incrementing GroupID (every entity spawned gets a new one) and SubGroupIDs are used to internally disable collision between parts of the same entity (e.g. torso and arm of a ragdoll). When two entities connect (and we want to disable their collision), we set the GroupID of both entities to the same ID (child ID becomes parent ID). If two entities have a different group filters they will not collide (as per the 2nd collision rule, see GroupFilterTable.h), so that's how they stop colliding. For your categorization you could either use a couple of bits from the (Sub)GroupID or perhaps you could move this information to the object layer. There are some holes in this system (e.g. if you spawn 4 billion entities you will roll over and 2 entities that are supposed to collide won't and if you connect 2 ragdolls of the same type together the torso of 1 will not collide with the arm of the other) but overall it works pretty well. I don't know if this will work for your case as that obviously depends on your setup.

**Reply by Avithohol** [reply]
> Yes, this is what the GroupFilter virtual interface is for. Each body has a CollisionGroup (which contains a GroupFilter reference). The Collision group has 64 bits of custom data and you can use that in the GroupFilter::CanCollide function to determine if two objects should collide. An example implementation is GroupFilterTable and an example of how to use it in the case of ragdolls can be found in RagdollSettings::DisableParentChildCollisions.

@jrouwe 
I know its old thread but i think it complements to it.

One use case I cannot solve with this, for example, 
spawning projectiles (dynamic bodies) inside a (character) capsule body, and let them fly out without colliding with the capsule.
(Basically player shooting missiles at other players)
These projectiles need to collide with other capsules (other players), but the one that spawned them, 
so filtering per group is not possible.

That is good example for where "per object" would greatly help.
Apology for my hypocrisy here, but I am moving from Bullet to Jolt, where there was `btCollisionObject::setIgnoreCollisionCheck(const btCollisionObject* co, bool ignoreCollisionCheck);` enabling this easily.

Is there a plan to to add this to Jolt? Or is there a solution for this use case?

**Reply by jrouwe** [reply]
It's definitively possible to do this, even with the standard `GroupFilterTable`. Each character gets its own unique GroupID, each rocket fired by a character inherits the GroupID of its character. The SubGroupID of the character and rockets could both be 0. This will give you what you describe above:

If the rocket hits another character, the GroupIDs will be different which means they will collide.
If the rocket hits the spawning character, the GroupIDs will be the same so we will look at the SubGroupIDs. Since they're both 0, they will not collide.

See:

https://github.com/jrouwe/JoltPhysics/blob/2021f6aa05f47ebd509e0803c2c8315403596ba1/Jolt/Physics/Collision/GroupFilterTable.h#L14-L32

But you can implement your own `GroupFilter` class. You have 64 bits (`GroupID` and `SubGroupID`) that you can use to identify which object is colliding (e.g. by storing `BodyID` in `GroupID` or by splitting a pointer to your game entity across `GroupID` and `SubGroupID`). Your `GroupFilter` can contain a table that specifies what collides with what just like Bullet does.

**Reply by Avithohol** [reply]
@jrouwe you are right. Just tested, and it works. This set both subgroup to 0 was not obvious for the first sight.
Now I am a bit uncertain that if this is the right way to spawn missiles, but it works, and happy about it.

@DocAce's case is supported too. I guess this case is closed.
